src/tools.rs

```diff
@@ -267,3 +267,180 @@ pub fn decode_swdord(data: &[u8]) -> i64 {
         udw as i64
     }
 }
+
+/// Parse basic types from a raw buffer (&[u8]).
+///
+/// Provide functions to parser values of basic types from the buffer
+/// sequentially.
+pub struct RawBufReader<'a> {
+    off: usize,
+    data: &'a [u8],
+}
```

Comment:
This is confusing.  If I have something like
```
  let mut buf = Vec::<u8>::new();
  let raw = buf.as_slice();
```
Then, I pass `raw` to another function. So, I should pass a `&mut &[u8]`!!
And it will be available on every `&[u8]` variable, but we don't need it on every single one.

Reply:
> Then, I pass raw to another function. So, I should pass a &mut &[u8]!!

It's how `Read` [works](https://doc.rust-lang.org/std/io/trait.Read.html#impl-Read-for-%26%5Bu8%5D). Think of it as nothing more than an extension trait.

> And it will be available on every &[u8] variable, but we don't need it on every single one.

Yes, because you can attempt to decode a uword from every byte slice. That's similar to how you can read from every byte slice.

I am not saying it has to be a trait, though it's probably not a bad choice. What I am saying is that we should not store an additional offset when it's already implied in the slice anyway.

Reply:
In our previous offline discussions, I agreed with this idea. But I regret it now.  Commonly, we need to know the current offset to do either relative offset or as a part of the cached information that allows quick data finding.

Reply:
> Commonly, we need to know the current offset to do either relative offset or as a part of the cached information that allows quick data finding. Even I don't use this feature in this patch; it will be used once we start changing the existing code.

Can you please provide a concrete example of what you are talking about? It's hard for me to agree with or refute hypothetical scenarios that at this point exist only in your head.

Reply:
If you check `debug_info::DIEIter`, you will see it has a `seek_to_sibling()` method to skip a DIE and its descendants quickly.  It seeks to an offset value that starts from the beginning of a compile unit. 

Some DIEs have a `sibling_offset` attribute, which is used to skip the whole subtree quickly.  `debug_info_parse_symbols_cu()` is an example of using this attribute to speed up parsing DWARF debug info.

Reply:
So we need to search backwards, is that what you are saying? Otherwise I fail to see the problem you are trying to point out, sorry.

It's also not clear to me at all what this functionality has to do with the problem at hand. Unless you are saying that it has to go into the trait (or be part of your RawBufReader)? Isn't the goal of the work to get rid of utility functions working on byte buffers that duplicate offset additions between caller and callee in error-prone ways? I am not suggesting that we group every custom parsing/reading functionality in this crate into one type/trait/whatever unit. On the contrary, we very much want ELF specific parsing co-located with ELF specific parsing and not club it together with, say, DWARF stuff. My ask was about the reduction of redundancy and error prone code.

Reply:
No, it seeks forwards, and we only have an offset (`sibling_offset`) from the beginning of the containing CU (compile unit; the whole buffer). The offset is not relative to the current position.

The goal here is to avoid error-prone code that maintains offsets and passes sizes everywhere.  `RawBufReader` will maintain the offsets for its users.  The issue that duplicates offset additions between callers and callees can also be removed by passing references of `RawBufReader`s to `callees`.

Reply:
> No, it seeks forwards, and we only have an offset (`sibling_offset`) from the beginning of the containing CU (compile unit; the whole buffer). The offset is not relative to the current position.

Okay. So then you'd pass in a slice to the beginning of the containing CU, do a `seek` or whatever we call it for `sibling_offset` bytes and you are there. It's not like the original slice is necessarily made unavailable in the process, so you can refer back to it to go to the next sibling later on. And again, I don't understand what connection this functionality has to what was originally identified as the problem.

> The goal here is to avoid error-prone code that maintains offsets and passes sizes everywhere. RawBufReader will maintain the offsets for its users. The issue that duplicates offset additions between callers and callees can also be removed by passing references of RawBufReaders to callees.

Aha. It's still not clear to me how that cannot be accomplished with what I proposed, which I argue is reduced to the minimum without any unnecessary duplication of data (`off` member is implied), easier to migrate to (no need to scatter `RawBufReader` objects across the code base, all you do is `use` a trait; no need to instantiate a new object just to reader a dword or similar), and makes better use of Rust's features to reduce the amount of accidental complexity. In short: I don't see what additional benefit `RawBufReader` provides.

Reply:
Just like what we discussed offline. I use the same case to explain why we need an offset.

As you suggested, when we need to seek a location from the beginning of the buffer, the code would look like this (with a slice, without an offset). (I hope I don't misunderstand you.)

      fn parse_with_slice(raw: &mut &[u8]) -> Result<(), Error> {
            let saved_raw = *raw;
       
           while ... {
               raw.read_word();
                ........
                let offset = raw.read_lib128()?; // `offset` is from the beginning of the buffer
                *raw = saved_raw;  // need to update it manually.
                raw.seek_to(offset)?;
                raw.read_leb123()?;
                ......
           }
      }

The code would look like this with my implementation (with a slice and offset).

      fn parse_with_rawbufreader(raw: &mut RawBufReader) -> Result<(), Error> {
            while ... {
                raw.read_word()?;
                ......
                let offset = raw.read_lib128()?; // `offset` is from the beginning of the buffer
                raw.seek_to(offset)?;
                raw.read_leb123()?;
                ......
            }
      }

It is straightforward with my implementation.

The trait is amazing. However, it is vague. You import a trait from another module, and suddenly the `&[u8]`  type has special power from the module that implements the trait. And people have no idea that the A module implemented these methods until you grep the whole repository.

Reply:
Well, you are clearly omitting the fact that you have *undocumented* and *implicit* knowledge here that `raw`'s internal slice points to the very beginning of what offsets are relative to, otherwise your logic won't work. Because what does `raw.seek_to(offset)` seek to if it does not? Some random garbage somewhere. Again, now you require that the provided `RawBufReader` instance is to be created anew for each function, basically, but that requirement is nowhere made explicit. It's dependent on the implementation of the function (if it happens to need to seek around it matters, otherwise it doesn't). That is horrible API design if you ask me and the approach that I've laid out makes that fact very explicit, while being more minimal, and requiring less code, etc., as I've mentioned before. You won't have to adjust callers based on a changing implementation.

That being said, I've laid out my case multiple times and I am tired of wasting time making arguments when they can be brushed off with what amounts to "I don't like it". So go ahead.

Reply:
I explained it yesterday in our offline discussion.  It is a contract between a caller and the callee. Are you going to complain `std::io::Seek` doesn't work? Your solution doesn't change it too, and you still need another copy that satisfies the needs of the callee.

Reply:
> It is a contract between a caller and the callee.

Yes, and it is unnecessary and not enforced at all, making it a very error prone approach, as I said earlier.

> Are you going to complain `std::io::Seek` doesn't work just because it doesn't remove the contract between callers and callees?

No, you are conflating issues.

> Your solution doesn't change it too, and you still need another copy that satisfies the needs of the callee.

Yes it does. Please read and understand [this comment](https://github.com/libbpf/blazesym/pull/36#discussion_r1061730182).